test/python/utils/test_deprecation.py

```diff
@@ -0,0 +1,464 @@
+                """\
+
+                .. deprecated:: 9.99_pending
+                  TestDeprecationDecorators.test_deprecate_arguments_message.<locals>.my_func \
+keyword argument old_arg1 is deprecated and replaced with new_arg1.
+
+                .. deprecated:: 9.99_pending
+                  TestDeprecationDecorators.test_deprecate_arguments_message.<locals>.my_func \
+keyword argument old_arg2 is deprecated and will in the future be removed.
+                """
+            ),
+        )
+
+    def test_deprecate_function_docstring(self) -> None:
+        """Test that `@deprecate_function` adds the correct message to the docstring."""
+
+        @deprecate_function("Stop using my_func!", since="9.99")
+        def my_func() -> None:
+            pass
+
+        self.assertEqual(
+            my_func.__doc__,
+            dedent(
+                """\
+
+                .. deprecated:: 9.99
+                  Stop using my_func!
+                """
+            ),
+        )
+
+    def test_deprecate_arguments_runtime_warning(self) -> None:
+        """Test that `@deprecate_arguments` warns whenever the arguments are used.
+
+        Also check that old arguments are passed in as their new alias.
+        """
+
+        @deprecate_arguments({"arg1": None, "arg2": "new_arg2"}, since="9.99")
+        def my_func(*, arg1: str = "a", new_arg2: str) -> None:
+            del arg1
+            assert new_arg2 == "z"
+
+        my_func(new_arg2="z")  # No warnings if no deprecated args used.
+        with self.assertWarns(DeprecationWarning):
+            my_func(arg1="a", new_arg2="z")
+        with self.assertWarns(DeprecationWarning):
+            # `arg2` should be converted into `new_arg2`.
+            my_func(arg2="z")  # pylint: disable=missing-kwoa
+
+    def test_deprecate_function_runtime_warning(self) -> None:
+        """Test that `@deprecate_function` warns whenever the function is used."""
+
+        @deprecate_function("Stop using my_func!", since="9.99")
+        def my_func() -> None:
+            pass
+
+        with self.assertWarns(DeprecationWarning):
+            my_func()
+
+
+class DeprecationExtensionTest(QiskitTestCase):
+    """Test that we correctly insert the deprecation directive at the right location.
+
+    These test cases were manually created by changing the docstring of a function in
+    ``qiskit-ibmq-provider`` and building the docs to ensure that everything rendered correctly.
+    (Any Qiskit repo will do.) To get the ``expected`` lines, you can add this extension to
+    ``conf.py``:
+
+        def print_docstring(app, what, name, obj, options, lines):
+            if name != "FILL ME IN WITH QISKIT.MODULE.FUNCTION_NAME":
+                return
+            print("HERE")
+            print(obj.__doc__)
+
+        def setup(app):
+            app.connect('autodoc-process-docstring', print_docstring)
+
+    Then, choose a function and modify it's docstring. Update the extension with that function's
```

Comment:
it couldn't possibly matter less since this isn't public-facing, but "it's" -> "its".  I think it might be our most common grammatical mistake across our docs haha